rfcs/2022-02-23-9481-config-schema.md

```diff
@@ -0,0 +1,823 @@
+implementors are fully documented: the type itself, their subfields, and so on. It would then become
+extremely hard for developers to add new fields or types in configuration that weren't documented
+upon cutting a new Vector release.
+
+#### 2. `Configurable` trait as a vehicle for encapsulating all facets of configuration
+
+As discussed in point 1, the `Configurable` trait is meant to provide a common interface for
+configuration types, and the types used within those types, such as types from the standard library
+or third-party crates, such that they can describe their "shape", value constraints, and any other
+relevant metadata. Below is an abbreviated version of the `Configurable` trait, along with
+supporting types:
+
+```rust
+/// The shape of the field.
+///
+/// This maps similiar to the concept of JSON's data types, where types are generalized and have
+/// generalized representations. This allows us to provide general-but-relevant mappings to core
+/// types, such as integers and strings and so on, while providing escape hatches for customized
+/// types that may be encoded and decoded via "normal" types but otherwise have specific rules or
+/// requirements.
+///
+/// Additionally, the shape of a field can encode some basic properties about the field to which it
+/// is attached. For example, numbers can be bounded on or the lower or upper end, while strings
+/// could define a minimum length, or even an allowed pattern via regular expressions.
+///
+/// In this way, they describe a more complete shape of the field than simply the data type alone.
+#[derive(Clone)]
+pub enum Shape {
+    Null,
+    Boolean,
+    String(StringShape),
+    Number(NumberShape),
+    Array(ArrayShape),
+    Map(MapShape),
+    Composite(Vec<Shape>),
+}
+
+#[derive(Clone, Default)]
+pub struct StringShape {
+    minimum_length: Option<usize>,
+    maximum_length: Option<usize>,
+    allowed_pattern: Option<&'static str>,
+}
+
+#[derive(Clone)]
+pub enum NumberShape {
+    Unsigned {
+        effective_lower_bound: u128,
+        effective_upper_bound: u128,
+    },
+    Signed {
+        effective_lower_bound: i128,
+        effective_upper_bound: i128,
+    },
+    FloatingPoint {
+        effective_lower_bound: f64,
+        effective_upper_bound: f64,
+    }
+}
+
+#[derive(Clone)]
+pub struct ArrayShape {
+    element_shape: Box<Shape>,
+    minimum_length: Option<usize>,
+    maximum_length: Option<usize>,
+}
+
+#[derive(Clone)]
+pub enum MapShape {
+    Fixed(HashMap<&'static str, Shape>),
+    Dynamic(Box<Shape>),
+}
+
+pub struct Field {
+    name: &'static str,
+    description: &'static str,
+    shape: Shape,
+    fields: Vec<Field>,
+    metadata: Option<Vec<Metadata>>,
+}
+
+pub enum TypedMetadata<T: Serialize> {
+    DefaultValue(T),
+    CustomAttribute(String, String),
+}
+
+pub enum Metadata {
+    DefaultValue(Value),
+    CustomAttribute(String, String),
+}
```

Comment:
Since it would be nonsense for a single type to have multiple default values (I think), would these be better described as `struct`s:
```suggestion
pub struct TypedMetadata<T: Serialize> {
    default: Option<T>,
    attributes: Vec<(String, String)>,
}

pub enum Metadata {
    default: Option<Value>,
    attributes: Vec<(String, String)>,
}
```
Then every type could return a (non-optional) metadata instead of an optional array of these enums.

Reply:
Yeah, this might be easier.

I was mostly trying to make the code generation a little simpler by allowing the procedural macro to build up a vector of metadata entries as it parsed, vs having to know up front exactly what the structure of `Metadata` is so that it could generate the correct struct initializer.

Thinking it through, though, and experimenting in the playground, I think what I could do is having the macro generate the field initializer lines -- i.e. `default: Some(...)` -- and then use the default splat initializer syntax to fill in the rest.

Something like this:

```rust
let metadata = Metadata {
        foo: Some(String::from("wee")),
        quux: true,
        ..Default::default()
    };
```

This does the right thing even if you've actually specified all the possible fields, so it would actually be just as much of a blind "map this helper attribute to this metadata struct field and chuck it into a vector" approach, and just as forwards-compatible.

I'll make this change.

Reply:
Did just this in c9b543cd0 and 409e4042a.

As a side benefit, since `serde_json::Value` is `Serialize`, I realized we could just drop the typed vs untyped metadata structs, and use a single typed one, where we'll just go from `Metadata<SomeConcreteType>` to `Metadata<Value>`.